sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShuffleCoalesceExec.scala

```diff
@@ -0,0 +1,207 @@
+  private[this] var numRowsInBatch: Int = 0
+  private[this] var batchByteSize: Long = 0L
+
+  override def hasNext: Boolean = {
+    withResource(new MetricRange(totalTimeMetric)) { _ =>
+      bufferNextBatch()
+      numTablesInBatch > 0
+    }
+  }
+
+  override def next(): ColumnarBatch = {
+    withResource(new MetricRange(totalTimeMetric)) { _ =>
+      if (!hasNext) {
+        throw new NoSuchElementException("No more columnar batches")
+      }
+      concatenateBatch()
+    }
+  }
+
+  private def batchIterHasNext: Boolean = {
+    withResource(new MetricRange(collectTimeMetric)) { _ =>
+      batchIter.hasNext
+    }
+  }
+
+  private def batchIterNext(): ColumnarBatch = {
+    withResource(new MetricRange(collectTimeMetric)) { _ =>
+      batchIter.next()
+    }
+  }
+
+  private def bufferNextBatch(): Unit = {
+    if (numTablesInBatch == serializedTables.size()) {
+      var batchCanGrow = batchByteSize < targetBatchByteSize
+      while (batchCanGrow && batchIterHasNext) {
+        closeOnExcept(batchIterNext()) { batch =>
+          inputBatchesMetric += 1
+          // don't bother tracking empty tables
+          if (batch.numRows > 0) {
+            inputRowsMetric += batch.numRows
+            val tableColumn = batch.column(0).asInstanceOf[SerializedTableColumn]
+            batchCanGrow = canAddToBatch(tableColumn.header)
+            serializedTables.addLast(tableColumn)
+            // always add the first table to the batch even if its beyond the target limits
+            if (batchCanGrow || numTablesInBatch == 0) {
+              numTablesInBatch += 1
+              numRowsInBatch += tableColumn.header.getNumRows
+              batchByteSize += tableColumn.header.getDataLen
+            }
+          } else {
+            batch.close()
+          }
+        }
+      }
+    }
+  }
+
+  private def canAddToBatch(nextTable: SerializedTableHeader): Boolean = {
+    if (batchByteSize + nextTable.getDataLen > targetBatchByteSize) {
+      return false
+    }
+    if (numRowsInBatch.toLong + nextTable.getNumRows > Integer.MAX_VALUE) {
+      return false
+    }
+    true
+  }
+
+  private def concatenateBatch(): ColumnarBatch = {
+    // about to start using the GPU in this task
+    GpuSemaphore.acquireIfNecessary(TaskContext.get())
+
+    val firstHeader = serializedTables.peekFirst().header
+    val batch = withResource(new MetricRange(concatTimeMetric)) { _ =>
+      if (firstHeader.getNumColumns == 0) {
+        (0 until numTablesInBatch).foreach(_ => serializedTables.removeFirst())
+        new ColumnarBatch(Array.empty, numRowsInBatch)
+      } else {
+        concatenateTablesBatch()
+      }
+    }
+
+    outputBatchesMetric += 1
+    outputRowsMetric += batch.numRows
+
+    // update the stats for the next batch in progress
+    numTablesInBatch = serializedTables.size
+    batchByteSize = 0
+    numRowsInBatch = 0
+    if (numTablesInBatch > 0) {
+      require(numTablesInBatch == 1)
```

Comment:
Ok this one confused me a bit, but I think I understand. `serializedTables.size - numTablesInBatch <= 1`, given the logic in `bufferNextBatch`. Should we add a string message to `require` in case this relationship changes? Or maybe a comment.